<h1 align=center><font size = 5>Peer-Graded Assignment</font></h1>

## A. Build a baseline model

<h3>Objective for this Section</h3>    
<h5>Use the Keras library to build a neural network with the following:</h5>

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

Let's first load required libraries:

In [1]:
import pandas as pd
import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense

### About dataset

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

Let's download the data and read it into a <em>pandas</em> dataframe.

In [2]:
# concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.

### Let's look at how many data points we have

In [3]:
concrete_data.shape

(1030, 9)

### Checking dataset for any missing values

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.


### 1. Train Test Split

Predictors Target Split

In [6]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( predictors, target, test_size=0.3, random_state=4)
print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

Train set: (721, 8) (721,)
Test set: (309, 8) (309,)


Let's save the number of predictors to *n_cols* since we will need this number when building our network.

In [8]:
n_cols = X_train.shape[1]

### 2. Train the model on the training data using 50 epochs.

Build Neural Network Model

In [9]:
def my_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# - One hidden layer of 10 nodes, and a ReLU activation function

### 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.


Train and Test Model using the *fit* and *predict* methods. 
We will train the model for 50 epochs.

In [10]:
# build the model
model = my_model()

# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=2)

# evaluate the model
scores = model.evaluate(X_test, y_test, verbose=0)

print("Mean Squared Error: %.2f"%scores)

Epoch 1/50
23/23 - 1s - loss: 238901.6094 - val_loss: 146889.9688 - 572ms/epoch - 25ms/step
Epoch 2/50
23/23 - 0s - loss: 91317.5234 - val_loss: 45452.9336 - 47ms/epoch - 2ms/step
Epoch 3/50
23/23 - 0s - loss: 25068.2949 - val_loss: 11325.0938 - 59ms/epoch - 3ms/step
Epoch 4/50
23/23 - 0s - loss: 7961.9824 - val_loss: 5755.8398 - 48ms/epoch - 2ms/step
Epoch 5/50
23/23 - 0s - loss: 6028.8164 - val_loss: 5268.2314 - 66ms/epoch - 3ms/step
Epoch 6/50
23/23 - 0s - loss: 5704.2148 - val_loss: 4966.4756 - 59ms/epoch - 3ms/step
Epoch 7/50
23/23 - 0s - loss: 5343.6509 - val_loss: 4693.4321 - 57ms/epoch - 2ms/step
Epoch 8/50
23/23 - 0s - loss: 5023.5938 - val_loss: 4401.9014 - 57ms/epoch - 2ms/step
Epoch 9/50
23/23 - 0s - loss: 4706.9541 - val_loss: 4126.2588 - 57ms/epoch - 2ms/step
Epoch 10/50
23/23 - 0s - loss: 4400.3501 - val_loss: 3870.1553 - 54ms/epoch - 2ms/step
Epoch 11/50
23/23 - 0s - loss: 4114.6348 - val_loss: 3639.5271 - 49ms/epoch - 2ms/step
Epoch 12/50
23/23 - 0s - loss: 3851.3020 -

### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [11]:
for i in range(1, 50):
    # build the model
    model = my_model()

    # fit the model
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=2)

    # evaluate the model
    score = model.evaluate(X_test, y_test, verbose=0)
    scores = np.append(scores, score)
    print("Iteration: ", i)
MSE = scores
MSE


Epoch 1/50
23/23 - 0s - loss: 1111.1124 - val_loss: 954.5981 - 364ms/epoch - 16ms/step
Epoch 2/50
23/23 - 0s - loss: 782.1798 - val_loss: 766.6254 - 44ms/epoch - 2ms/step
Epoch 3/50
23/23 - 0s - loss: 602.2813 - val_loss: 612.8683 - 55ms/epoch - 2ms/step
Epoch 4/50
23/23 - 0s - loss: 468.3327 - val_loss: 497.2317 - 40ms/epoch - 2ms/step
Epoch 5/50
23/23 - 0s - loss: 356.8408 - val_loss: 401.9052 - 43ms/epoch - 2ms/step
Epoch 6/50
23/23 - 0s - loss: 293.5402 - val_loss: 349.8746 - 52ms/epoch - 2ms/step
Epoch 7/50
23/23 - 0s - loss: 255.4214 - val_loss: 320.1862 - 45ms/epoch - 2ms/step
Epoch 8/50
23/23 - 0s - loss: 232.4127 - val_loss: 291.7062 - 54ms/epoch - 2ms/step
Epoch 9/50
23/23 - 0s - loss: 207.6278 - val_loss: 273.1529 - 49ms/epoch - 2ms/step
Epoch 10/50
23/23 - 0s - loss: 193.8219 - val_loss: 255.9992 - 45ms/epoch - 2ms/step
Epoch 11/50
23/23 - 0s - loss: 181.2216 - val_loss: 243.1022 - 46ms/epoch - 2ms/step
Epoch 12/50
23/23 - 0s - loss: 169.9687 - val_loss: 247.2576 - 45ms/epo

array([ 506.38208008,  114.45848846,  212.45951843,  118.13648224,
        126.4712677 ,   92.63408661,  493.86196899, 1036.61425781,
        136.23236084,  126.73427582, 1561.36828613,  143.80206299,
        252.86857605,  642.56652832,  397.30264282,  164.64932251,
        419.52807617,  393.47491455, 1444.99255371,  131.73167419,
        409.09747314,  172.18191528,  190.87080383,  330.24304199,
        177.58604431,  228.60671997, 1450.1105957 ,  353.22384644,
        266.73822021,  175.2848053 ,  152.4539032 ,   95.61823273,
        949.90289307,  735.67749023,  268.08969116,  149.77510071,
        120.17108154,  141.75047302,  306.69485474,  232.10992432,
        151.77511597,  793.71313477,  409.29946899,  154.39813232,
       1087.74523926,  190.12940979,  136.98497009,   78.88006592,
        151.47099304,  250.59159851])

### 5. Report the mean and the standard deviation of the mean squared errors.

In [12]:
print("Mean of the mean squared error: ",MSE.mean())
print("Standard deviation of the mean squared error: ",MSE.std())

Mean of the mean squared error:  376.5488932800293
Standard deviation of the mean squared error:  370.30420986100995


## B. Normalize the data



<h3>Objective for this Section</h3> 

- Repeat Part A but use a normalized version of the data. 
- Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.
- How does the mean of the mean squared errors compare to that from Step A?

Normalization

In [13]:
predictors = (predictors - predictors.mean())/predictors.std()
predictors.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [14]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( predictors, target, test_size=0.3, random_state=4)
print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

Train set: (721, 8) (721,)
Test set: (309, 8) (309,)


Let's save the number of predictors to *n_cols* since we will need this number when building our network.

In [15]:
n_cols = X_train.shape[1]

### 2. Train the model on the training data using 50 epochs.

In [16]:
def my_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# - One hidden layer of 10 nodes, and a ReLU activation function

### 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.


### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

Train and Test Model using the *fit* and *predict* methods. 
We will train the model for 50 epochs.

In [17]:
# build the model
model = my_model()

# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=2)

# evaluate the model
scores_B = model.evaluate(X_test, y_test, verbose=0)

print("Mean Squared Error: %.2f"%scores_B)

Epoch 1/50
23/23 - 1s - loss: 1532.9846 - val_loss: 1482.6642 - 523ms/epoch - 23ms/step
Epoch 2/50
23/23 - 0s - loss: 1517.0045 - val_loss: 1467.9857 - 68ms/epoch - 3ms/step
Epoch 3/50
23/23 - 0s - loss: 1500.8392 - val_loss: 1452.8982 - 69ms/epoch - 3ms/step
Epoch 4/50
23/23 - 0s - loss: 1483.9818 - val_loss: 1437.2421 - 84ms/epoch - 4ms/step
Epoch 5/50
23/23 - 0s - loss: 1466.3630 - val_loss: 1420.9258 - 76ms/epoch - 3ms/step
Epoch 6/50
23/23 - 0s - loss: 1447.8018 - val_loss: 1403.4454 - 84ms/epoch - 4ms/step
Epoch 7/50
23/23 - 0s - loss: 1427.7738 - val_loss: 1385.0573 - 84ms/epoch - 4ms/step
Epoch 8/50
23/23 - 0s - loss: 1406.8496 - val_loss: 1365.5221 - 79ms/epoch - 3ms/step
Epoch 9/50
23/23 - 0s - loss: 1384.7971 - val_loss: 1344.5098 - 89ms/epoch - 4ms/step
Epoch 10/50
23/23 - 0s - loss: 1361.3931 - val_loss: 1322.8116 - 66ms/epoch - 3ms/step
Epoch 11/50
23/23 - 0s - loss: 1336.8912 - val_loss: 1299.8717 - 66ms/epoch - 3ms/step
Epoch 12/50
23/23 - 0s - loss: 1311.3934 - val_los

In [18]:
for i in range(1, 50):
    # build the model
    model = my_model()

    # fit the model
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=2)

    # evaluate the model
    score_B = model.evaluate(X_test, y_test, verbose=0)
    scores_B = np.append(scores_B, score_B)
    print("Iteration: ", i)
MSE = scores_B
MSE


Epoch 1/50
23/23 - 1s - loss: 1597.3895 - val_loss: 1533.0813 - 529ms/epoch - 23ms/step
Epoch 2/50
23/23 - 0s - loss: 1577.8544 - val_loss: 1516.3893 - 66ms/epoch - 3ms/step
Epoch 3/50
23/23 - 0s - loss: 1558.9376 - val_loss: 1499.5452 - 65ms/epoch - 3ms/step
Epoch 4/50
23/23 - 0s - loss: 1539.6060 - val_loss: 1482.8910 - 77ms/epoch - 3ms/step
Epoch 5/50
23/23 - 0s - loss: 1520.5562 - val_loss: 1465.5850 - 63ms/epoch - 3ms/step
Epoch 6/50
23/23 - 0s - loss: 1500.7567 - val_loss: 1447.6736 - 65ms/epoch - 3ms/step
Epoch 7/50
23/23 - 0s - loss: 1480.3722 - val_loss: 1429.2936 - 65ms/epoch - 3ms/step
Epoch 8/50
23/23 - 0s - loss: 1459.4332 - val_loss: 1410.0239 - 61ms/epoch - 3ms/step
Epoch 9/50
23/23 - 0s - loss: 1437.4774 - val_loss: 1390.5830 - 84ms/epoch - 4ms/step
Epoch 10/50
23/23 - 0s - loss: 1414.9896 - val_loss: 1370.0425 - 89ms/epoch - 4ms/step
Epoch 11/50
23/23 - 0s - loss: 1391.6727 - val_loss: 1348.3273 - 62ms/epoch - 3ms/step
Epoch 12/50
23/23 - 0s - loss: 1367.0759 - val_los

array([296.32849121, 420.31195068, 375.55938721, 374.1423645 ,
       349.8656311 , 380.09002686, 441.63763428, 378.71813965,
       338.33792114, 409.04910278, 340.66864014, 388.28939819,
       546.77661133, 518.203125  , 381.81185913, 303.03787231,
       388.78378296, 617.02929688, 485.00524902, 268.48950195,
       568.41345215, 378.87973022, 317.20962524, 437.54232788,
       349.97885132, 429.60073853, 327.99713135, 507.17312622,
       615.94433594, 438.10043335, 289.70864868, 490.38531494,
       449.59365845, 400.67922974, 316.74307251, 557.15118408,
       312.71343994, 315.3963623 , 488.22491455, 487.21166992,
       295.23681641, 371.54620361, 412.18933105, 541.86633301,
       413.16088867, 321.77792358, 485.56027222, 269.31842041,
       388.45248413, 361.45678711])

### 5. Report the mean and the standard deviation of the mean squared errors.

In [19]:
print("Mean of the mean squared error: ",MSE.mean())
print("Standard deviation of the mean squared error: ",MSE.std())

Mean of the mean squared error:  406.8269738769531
Standard deviation of the mean squared error:  88.7572078395569


- Mean of the mean squared error A:  376.5488932800293
- Standard deviation of the mean squared error A:  370.30420986100995

- Mean of the mean squared error B:  406.8269738769531
- Standard deviation of the mean squared error B:  88.7572078395569

### Although the mean of the mean squared error increased slightly by ±30, the standard deviation has decreased significantly by ±282. This means that the normalization of the predictors has decreased the deviation and variation of the average mean squared error.

## C. Increate the number of epochs


- Repeat Part B but use 100 epochs this time for training.
- How does the mean of the mean squared errors compare to that from Step B?

### 2. Train the model on the training data using 100 epochs.

### 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.


### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

Train and Test Model using the *fit* and *predict* methods. 
We will train the model for 100 epochs.

In [20]:
# build the model
model = my_model()

# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=100, verbose=2)

# evaluate the model
scores_C = model.evaluate(X_test, y_test, verbose=0)

print("Mean Squared Error: %.2f"%scores_C)

Epoch 1/100
23/23 - 0s - loss: 1613.1362 - val_loss: 1549.3822 - 480ms/epoch - 21ms/step
Epoch 2/100
23/23 - 0s - loss: 1595.9542 - val_loss: 1534.0988 - 59ms/epoch - 3ms/step
Epoch 3/100
23/23 - 0s - loss: 1579.1302 - val_loss: 1519.1315 - 69ms/epoch - 3ms/step
Epoch 4/100
23/23 - 0s - loss: 1562.6921 - val_loss: 1504.2280 - 84ms/epoch - 4ms/step
Epoch 5/100
23/23 - 0s - loss: 1546.2562 - val_loss: 1489.1593 - 91ms/epoch - 4ms/step
Epoch 6/100
23/23 - 0s - loss: 1529.6284 - val_loss: 1473.6283 - 58ms/epoch - 3ms/step
Epoch 7/100
23/23 - 0s - loss: 1512.7268 - val_loss: 1457.8046 - 51ms/epoch - 2ms/step
Epoch 8/100
23/23 - 0s - loss: 1495.4880 - val_loss: 1441.5594 - 66ms/epoch - 3ms/step
Epoch 9/100
23/23 - 0s - loss: 1477.5627 - val_loss: 1424.9998 - 80ms/epoch - 3ms/step
Epoch 10/100
23/23 - 0s - loss: 1459.3815 - val_loss: 1407.3376 - 83ms/epoch - 4ms/step
Epoch 11/100
23/23 - 0s - loss: 1440.3136 - val_loss: 1389.3718 - 105ms/epoch - 5ms/step
Epoch 12/100
23/23 - 0s - loss: 1420.6

In [21]:
for i in range(1, 50):
    # build the model
    model = my_model()

    # fit the model
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=100, verbose=2)

    # evaluate the model
    score_C = model.evaluate(X_test, y_test, verbose=0)
    scores_C = np.append(scores_C, score_C)
    print("Iteration: ", i)
MSE = scores_C
MSE


Epoch 1/100
23/23 - 1s - loss: 1507.3378 - val_loss: 1449.5806 - 605ms/epoch - 26ms/step
Epoch 2/100
23/23 - 0s - loss: 1490.7522 - val_loss: 1434.3269 - 85ms/epoch - 4ms/step
Epoch 3/100
23/23 - 0s - loss: 1473.4489 - val_loss: 1417.3195 - 59ms/epoch - 3ms/step
Epoch 4/100
23/23 - 0s - loss: 1454.4260 - val_loss: 1399.3511 - 68ms/epoch - 3ms/step
Epoch 5/100
23/23 - 0s - loss: 1434.0494 - val_loss: 1379.9956 - 72ms/epoch - 3ms/step
Epoch 6/100
23/23 - 0s - loss: 1412.1891 - val_loss: 1359.0536 - 75ms/epoch - 3ms/step
Epoch 7/100
23/23 - 0s - loss: 1389.2303 - val_loss: 1336.3812 - 86ms/epoch - 4ms/step
Epoch 8/100
23/23 - 0s - loss: 1364.0563 - val_loss: 1313.0455 - 75ms/epoch - 3ms/step
Epoch 9/100
23/23 - 0s - loss: 1337.7761 - val_loss: 1288.1622 - 69ms/epoch - 3ms/step
Epoch 10/100
23/23 - 0s - loss: 1310.2570 - val_loss: 1261.7168 - 79ms/epoch - 3ms/step
Epoch 11/100
23/23 - 0s - loss: 1281.5090 - val_loss: 1234.0842 - 54ms/epoch - 2ms/step
Epoch 12/100
23/23 - 0s - loss: 1251.52

array([203.16578674, 171.57406616, 192.84283447, 180.49008179,
       211.22741699, 180.89530945, 175.06419373, 172.1489563 ,
       176.26928711, 171.91311646, 187.65119934, 187.30317688,
       175.34135437, 166.6010437 , 174.91360474, 179.88609314,
       175.85133362, 175.41888428, 171.02619934, 212.40643311,
       175.88459778, 166.98773193, 173.28903198, 155.48684692,
       159.24728394, 176.65979004, 167.98684692, 217.42472839,
       180.61065674, 173.84313965, 178.34675598, 177.46864319,
       267.28607178, 158.59510803, 199.27075195, 185.81718445,
       194.4210968 , 230.27932739, 164.90945435, 171.87554932,
       172.31204224, 166.73170471, 202.80342102, 177.74894714,
       211.81744385, 178.06747437, 185.91905212, 177.04202271,
       207.8572998 , 194.02171326])

### 5. Report the mean and the standard deviation of the mean squared errors.

In [22]:
print("Mean of the mean squared error: ",MSE.mean())
print("Standard deviation of the mean squared error: ",MSE.std())

Mean of the mean squared error:  183.84004180908204
Standard deviation of the mean squared error:  19.92773529156402


- Mean of the mean squared error B:  406.8269738769531
- Standard deviation of the mean squared error B:  88.7572078395569
- Mean of the mean squared error C:  183.84004180908204
- Standard deviation of the mean squared error C:  19.92773529156402

### The mean and standard deviation of the mean squared error has decreased significantly by ±217 and ±68 respectively. Using the methods in C means that the mean squared error loss tends to be less and more stable than the methods in B. However, it requires more computational resources since it took twice the time to run compared to methods in B as the epoch doubled in size.

## D. Increase the number of hidden layers

#### Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

- How does the mean of the mean squared errors compare to that from Step B?

### 2. Train the model on the training data using 50 epochs.

Build Neural Network Model

In [23]:
def my_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# - Three hidden layers of 10 nodes, and a ReLU activation function

### 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.


### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

Train and Test Model using the *fit* and *predict* methods. 
We will train the model for 50 epochs.

In [24]:
# build the model
model = my_model()

# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=2)

# evaluate the model
scores_D = model.evaluate(X_test, y_test, verbose=0)

print("Mean Squared Error: %.2f"%scores_D)

Epoch 1/50
23/23 - 1s - loss: 1572.0085 - val_loss: 1505.0764 - 650ms/epoch - 28ms/step
Epoch 2/50
23/23 - 0s - loss: 1547.2566 - val_loss: 1481.0531 - 59ms/epoch - 3ms/step
Epoch 3/50
23/23 - 0s - loss: 1515.1746 - val_loss: 1445.4375 - 57ms/epoch - 2ms/step
Epoch 4/50
23/23 - 0s - loss: 1465.9661 - val_loss: 1390.6910 - 55ms/epoch - 2ms/step
Epoch 5/50
23/23 - 0s - loss: 1389.7627 - val_loss: 1305.5875 - 61ms/epoch - 3ms/step
Epoch 6/50
23/23 - 0s - loss: 1272.9709 - val_loss: 1170.5308 - 57ms/epoch - 2ms/step
Epoch 7/50
23/23 - 0s - loss: 1098.9716 - val_loss: 982.6204 - 60ms/epoch - 3ms/step
Epoch 8/50
23/23 - 0s - loss: 881.6774 - val_loss: 763.0816 - 62ms/epoch - 3ms/step
Epoch 9/50
23/23 - 0s - loss: 654.4866 - val_loss: 555.2801 - 61ms/epoch - 3ms/step
Epoch 10/50
23/23 - 0s - loss: 465.0232 - val_loss: 403.6196 - 58ms/epoch - 3ms/step
Epoch 11/50
23/23 - 0s - loss: 341.4821 - val_loss: 306.5417 - 63ms/epoch - 3ms/step
Epoch 12/50
23/23 - 0s - loss: 267.9262 - val_loss: 258.408

In [25]:
for i in range(1, 50):
    # build the model
    model = my_model()

    # fit the model
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=2)

    # evaluate the model
    score_D = model.evaluate(X_test, y_test, verbose=0)
    scores_D = np.append(scores_D, score_D)
    print("Iteration: ", i)
MSE = scores_D
MSE


Epoch 1/50
23/23 - 0s - loss: 1589.0083 - val_loss: 1516.5238 - 386ms/epoch - 17ms/step
Epoch 2/50
23/23 - 0s - loss: 1559.3965 - val_loss: 1490.4495 - 50ms/epoch - 2ms/step
Epoch 3/50
23/23 - 0s - loss: 1528.6587 - val_loss: 1458.7299 - 50ms/epoch - 2ms/step
Epoch 4/50
23/23 - 0s - loss: 1485.8895 - val_loss: 1408.1381 - 49ms/epoch - 2ms/step
Epoch 5/50
23/23 - 0s - loss: 1413.5527 - val_loss: 1316.1331 - 54ms/epoch - 2ms/step
Epoch 6/50
23/23 - 0s - loss: 1282.2965 - val_loss: 1161.7996 - 54ms/epoch - 2ms/step
Epoch 7/50
23/23 - 0s - loss: 1080.7035 - val_loss: 953.2431 - 59ms/epoch - 3ms/step
Epoch 8/50
23/23 - 0s - loss: 837.6915 - val_loss: 722.9725 - 65ms/epoch - 3ms/step
Epoch 9/50
23/23 - 0s - loss: 607.6116 - val_loss: 537.6608 - 66ms/epoch - 3ms/step
Epoch 10/50
23/23 - 0s - loss: 448.4320 - val_loss: 417.8648 - 65ms/epoch - 3ms/step
Epoch 11/50
23/23 - 0s - loss: 351.9580 - val_loss: 344.5539 - 62ms/epoch - 3ms/step
Epoch 12/50
23/23 - 0s - loss: 294.4807 - val_loss: 290.421

array([146.18562317, 147.74745178, 139.66992188,  93.7428894 ,
       147.57281494, 149.47113037, 142.99520874, 146.70051575,
       141.42059326, 101.97275543, 149.1295929 , 147.65408325,
       145.11035156, 130.26704407, 109.78317261, 131.9181366 ,
       159.77487183, 106.71965027, 149.18258667, 123.25415802,
       147.96360779, 157.21253967, 145.60340881, 110.12501526,
       145.11016846, 140.10375977, 144.68544006, 106.07843781,
       140.08361816, 138.39855957, 144.64837646, 142.412323  ,
       137.44659424, 152.48960876, 145.14398193, 151.03083801,
       148.95516968, 145.50627136, 106.66931152, 142.26002502,
       152.80021667, 150.87487793, 152.54226685, 135.3585968 ,
       151.95428467, 145.4478302 , 142.93603516, 147.06561279,
       150.13450623, 148.66003418])

### 5. Report the mean and the standard deviation of the mean squared errors.

In [26]:
print("Mean of the mean squared error: ",MSE.mean())
print("Standard deviation of the mean squared error: ",MSE.std())

Mean of the mean squared error:  139.5994773864746
Standard deviation of the mean squared error:  15.429176658438593


- Mean of the mean squared error B:  406.8269738769531
- Standard deviation of the mean squared error B:  88.7572078395569
- Mean of the mean squared error D:  139.5994773864746
- Standard deviation of the mean squared error D:  15.429176658438593

### The mean and standard deviation of the mean squared error has decreased significantly by ±167 and ±73 respectively. Using the methods in D means that the mean squared error loss tends to be less and more stable than the methods in B. In addition, it requires less computational resources since it took less time to run compared to methods in B as the number of hidden layers increased.

## Conclusion

We can confidenlty say that normilizing the predictors and increasing the number of hidden layers in the model is the most efficient way to decrease the mean squared error loss of our model.

## This notebook was created by Thabo Magagula